trieste/utils/pareto.py

```diff
@@ -101,10 +110,161 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
         )
 
         range_ = tf.range(len_front + 1)[:, None]
-        lower = tf.concat([range_, tf.zeros_like(range_)], axis=-1)
-        upper = tf.concat([range_ + 1, pf_ext_idx[::-1, 1:][: pf_ext_idx[-1, 0]]], axis=-1)
+        lower_result = tf.concat([range_, tf.zeros_like(range_)], axis=-1)
+        upper_result = tf.concat(
+            [range_ + 1, pseudo_front_idx[::-1, 1:][: pseudo_front_idx[-1, 0]]], axis=-1
+        )
+
+        return BoundedVolumes(lower_result, upper_result)
+
+    @staticmethod
+    def _is_test_required(smaller: TensorType) -> TensorType:
+        idx_dom_augm = tf.reduce_any(smaller, axis=1)
+        is_dom_augm = tf.reduce_all(idx_dom_augm)
+
+        return is_dom_augm
+
+    def _divide_conquer_nd(
+        self, front: TensorType, threshold: TensorType | float = 0
+    ) -> TensorType:
+        # Divide and conquer strategy to compute the cells covering the non-dominated region.
+        # Generic version: works for an arbitrary number of objectives.
+
+        len_front, number_of_objectives = front.shape
+        jitter = gpflow.config.default_jitter()
+        lower_result = tf.zeros([0, number_of_objectives], dtype=tf.int32)
+        upper_result = tf.zeros([0, number_of_objectives], dtype=tf.int32)
 
-        return BoundedVolumes(lower, upper)
+        min_front = tf.reduce_min(front, axis=0, keepdims=True) - 1
+        max_front = tf.reduce_max(front, axis=0, keepdims=True) + 1
+        pseudo_front = tf.concat([min_front, front, max_front], axis=0)
+
+        pseudo_front_idx = tf.concat(
+            [
+                tf.zeros([1, number_of_objectives], dtype=tf.int32),
+                tf.argsort(front, axis=0) + 1,  # +1 as index zero is reserved for the ideal point
+                tf.ones([1, number_of_objectives], dtype=tf.int32) * len_front + 1,
+            ],
+            axis=0,
+        )
+
+        divide_conquer_cells = tf.stack(
+            [
+                tf.zeros(number_of_objectives, dtype=tf.int32),
+                (int(pseudo_front_idx.shape[0]) - 1)
+                * tf.ones(number_of_objectives, dtype=tf.int32),
+            ],
+            axis=0,
+        )[None]
+
+        total_size = tf.reduce_prod(max_front - min_front)
+
+        def while_body(
+            divide_conquer_cells: TensorType,
+            lower_result: TensorType,
+            upper_result: TensorType,
+        ) -> Tuple[TensorType, TensorType, TensorType]:
+            divide_conquer_cells = tf.unstack(divide_conquer_cells, axis=0)
+            cell = divide_conquer_cells[-1]
+            divide_conquer_cells = tf.cond(
```

Comment:
this also might not work as expected with `tf.function` (re-assigning to `divide_conquer_cells`), but should be as simple as using a new variable name?

Reply:
I see didn't know that. hmm. is it possible to fix all these reassignments using a new variable name? if so, I am thinking what name should those be.

Reply:
okay